stablehlo/dialect/TypeInference.cpp

```diff
@@ -2757,7 +2752,45 @@ LogicalResult inferTriangularSolveOp(
     return emitOptionalError(
         location, "Invalid transpose option value for triangular solve");
 
-  inferredReturnShapes.emplace_back(bType.cast<ShapedType>());
+  auto aBounds = encodingToBounds(aType.getEncoding()).vec();
+  auto bBounds = encodingToBounds(bType.getEncoding()).vec();
+  if (aBounds.empty() || bBounds.empty()) {
+    inferredReturnShapes.emplace_back(bType.cast<ShapedType>());
+    return success();
+  }
+  auto resultShape = bType.getShape().vec();
+  auto resultBounds = encodingToBounds(aType.getEncoding()).vec();
+  for (int64_t i = 0; i < aRank - 2; i++) {
+    auto inferredDimAndBoundOrErr = inferMergedDimAndBound(
+        location, i, aBatchDims[i], bBatchDims[i], aBounds[i], bBounds[i]);
+    if (failed(inferredDimAndBoundOrErr)) return failure();
+    resultShape[i] = (*inferredDimAndBoundOrErr).first;
+    resultBounds[i] = (*inferredDimAndBoundOrErr).second;
+  }
+  if (leftSide) {
+    resultShape[bRank - 2] = bType.getShape()[bRank - 2];
+    resultBounds[bRank - 2] = bBounds[bRank - 2];
+    if (isTransposeOrAdjoint) {
+      resultShape[bRank - 1] = aType.getShape()[aRank - 1];
+      resultBounds[bRank - 1] = aBounds[aRank - 1];
+    } else {
+      resultShape[bRank - 1] = aType.getShape()[aRank - 2];
+      resultBounds[bRank - 1] = aBounds[aRank - 2];
+    }
+  } else {
+    resultShape[bRank - 1] = bType.getShape()[bRank - 1];
+    resultBounds[bRank - 1] = bBounds[bRank - 1];
+    if (isTransposeOrAdjoint) {
+      resultShape[bRank - 2] = aType.getShape()[aRank - 2];
+      resultBounds[bRank - 2] = aBounds[aRank - 2];
+    } else {
+      resultShape[bRank - 2] = aType.getShape()[aRank - 1];
+      resultBounds[bRank - 2] = aBounds[aRank - 1];
+    }
+  }
```

Comment:
Also, there's a considerable degree of duplication between the branches that looks like it can be relatively easily avoided. If we decide to keep this code (rather than a simpler "inherit bounds from `bType`" strategy), let's discuss how to improve this.

Reply:
Let's inherit bounds from bType and leave more specific bound inference for future work.